### DISCUSSION TITLE
Resizing a texture asset at runtime

### DISCUSSION THREAD

**Discussion by NicusorN5** [stride3d/stride#1693] [Q&A]
Hello, I have a render target texture that I want to resize to the size of the back buffer at runtime (namely at a `Start()` method inside a `SyncScript`). I can't seem any method to do that.

Any ideas on how to do that?

Thanks

**Comment by NicusorN5** [comment]
I'm considering implementing a Resize method for the Texture class. Any other opinions?

**Reply by Eideren** [reply]
I'm not sure how feasible that is given that textures are always used as fixed size buffers, but I wont dissuade you from trying. 
@tebjan has more experience in this than I do so I'm sure he'll have some insight to share

**Reply by Ethereal77** [reply]
Usually resizing a texture implies creating a new texture with the new size and the same format, and optionally destroy the original one. As this is a "low-level" graphics thing that should be done consciously by the user, don't know if a `Resize()` method is the best way.
What would happen to the original texture when being resized?

Also, as textures are graphics resources, maybe the resize operation should be done by the graphics resource manager (I think in Stride that task is done by `GraphicsDevice`, but I'll need to check).

**Reply by Ethereal77** [reply]
If not in the graphics device, maybe in a static method like `Texture.NewResized(Texture t, int newWidth, int newHeight, ...)`?

**Reply by NicusorN5** [reply]
> What would happen to the original texture when being resized?

Obviously get replaced.

> maybe in a static method like Texture.NewResized(Texture t, int newWidth, int newHeight, ...)

I exactly DO NOT want that. I want to resize a existing texture, not clone then resize.

**Reply by Ethereal77** [reply]
That's not how resources work in graphics APIs. You can't replace a resource. You create a new one, copy the contents from the old one, then (if you want) destroy the old one.
That's why I ask what would happen to the original texture.

Also, it would not be a "clone then resize", but a "new resized one then clone".

**Reply by Ethereal77** [reply]
Also, when you say "resize" you mean simply altering the dimensions of the texture (width, height)? Or also stretching the contents to the new size?

**Reply by NicusorN5** [reply]
> You create a new one, copy the contents from the old one, then (if you want) destroy the old one.

By replacing I mean these steps, also destrying the old one and replacing the texture handles in the current texture instance.

**Reply by Ethereal77** [reply]
Well, then yes. It would be possible. But as `Texture` is just a thin wrapper over what low-level APIs offer with their texture and resource APIs, don't know if this behavior should be here or another place more appropriate for these high-level utils.

I think right now, `Texture` mimics pretty closely a Direct3D texture resource (as implemented in SharpDX for example).

**Reply by tebjan** [reply]
@Ethereal77 is correct, just dispose the old one and make a new one and use that. There are many methods in stride to create textures, depending on what features the texture should have and what it will be used for.

Can you explain more about your use case?

**Comment by NicusorN5** [comment]
> Can you explain more about your use case?

Very late reply (I even forgot this thread even existed), but I want to **resize a Render Target asset to have the same size as the back buffer**.

Update1: I also have realized that I've specified the use case in the initial post, so writing this was useless.

**Reply by NicusorN5** [reply]
*bump*

**Reply by Eideren** [reply]
To resize a texture you have to re-allocate the memory, it works like an array does in c#, if you want a bigger or smaller one, you create a new one and swap references.

You can technically just swap the internal buffer the texture points to, but there are a lot of gotchas here because a lot of our graphics layer wouldn't expect a `Texture` to ever get re-allocated from under them, materials for instance may have caching behavior which only sends back to gpu pointers to their textures data when they are swapped for another one.

So I can see two ways you can go about this, 
- Either you create a new texture and swap all references that pointed to the old one to this new one.
- Or you create a new type that does not inherit from texture, and create the api necessary to feed this new dynamic texture to whatever system you need.

**Reply by dotlogix** [reply]
Any progress on this I also need to resize images but cant find any resources how to do it :'(

**Reply by Idomeneas1970** [reply]
Try this (rendering to texture) to resize and reformat. 

```
        public static Texture Resize(this Texture texin, int width, int height,
            GraphicsContext GraphicsContext, 
            PixelFormat pixelformat= PixelFormat.R8G8B8A8_UNorm    )
        {
            //if(texin.Width==width && texin.Height == height) return texin;
            GraphicsDevice GraphicsDevice=texin.GraphicsDevice;
            Texture offlineTarget = Texture.New2D(GraphicsDevice, width, height,
                pixelformat, TextureFlags.ShaderResource | TextureFlags.RenderTarget);
            Texture depthBuffer = Texture.New2D(GraphicsDevice, width, height, 
                PixelFormat.D24_UNorm_S8_UInt , TextureFlags.DepthStencil);
            SpriteBatch spriteBatch = new SpriteBatch(GraphicsDevice);

            // render into texture
            GraphicsContext.CommandList.Clear(offlineTarget, new Color4(0, 0, 0, 0));
            GraphicsContext.CommandList.Clear(depthBuffer, DepthStencilClearOptions.DepthBuffer);
            GraphicsContext.CommandList.SetRenderTargetAndViewport(depthBuffer, offlineTarget);

            spriteBatch.Begin(GraphicsContext);
            spriteBatch.Draw(texin, new RectangleF(0, 0, width, height), null, Color.White, 0, Vector2.Zero);
            spriteBatch.End();

            // copy texture on screen
            GraphicsContext.CommandList.Clear(GraphicsDevice.Presenter.BackBuffer, Color.Black);
            GraphicsContext.CommandList.Clear(GraphicsDevice.Presenter.DepthStencilBuffer, DepthStencilClearOptions.DepthBuffer);
            GraphicsContext.CommandList.SetRenderTargetAndViewport(GraphicsDevice.Presenter.DepthStencilBuffer, GraphicsDevice.Presenter.BackBuffer);

            spriteBatch.Begin(GraphicsContext);
            spriteBatch.Draw(offlineTarget, new RectangleF(0, 0, width, height), null, Color.White, 0, Vector2.Zero);
            spriteBatch.End();
            return offlineTarget;
        }
```
